### DISCUSSION TITLE
how to set highlight for specific tree-sitter groups

### DISCUSSION THREAD

**Discussion by ExpandingMan** [neovim/neovim#24451] [Q&A] [answered]
I'm checking highlights with `TSHighlightUnderCursor`.  I have a highlight group for which it lists `@symbol`, `@variable`, in that order.  What I want is for `@variable` to have color 1, `@symbol` to have color 2 and syntax that lists *both* `@symbol` and `@variable` in this way to have color 2.  However, after everything I have tried, it always seems to inherit the color of `@variable`.  How do I change this?

I have tried, for example
```lua
vim.api.nvim_set_hl(0, "@variable", {fg=colors.foreground})
vim.api.nvim_set_hl(0, "@symbol", {fg=colors.orange})
```
and 
```lua
vim.api.nvim_set_hl(0, "@variable.symbol", {fg=colors.orange})
```
and variations thereof, but nothing has worked.

**Comment by clason** [comment, marked as answer]
Can't be done; last extmark (and these are extmarks) always wins.

**Reply by haolian9** [reply]
do you mind if i ask what makes a treesitter capture group be last?
the order is the same as the capture groups defined in the highlights.scm?

**Reply by clason** [reply]
Yes, and whatever file is last if they are concatenated.

**Comment by ExpandingMan** [comment]
That's unfortunate.  So it sounds like I should open an issue with the languages tree-sitter indicating that I do not believe the parsing gives sufficiently specific highlight groups.  Thanks.

**Comment by sasja-san** [comment]
I'm trying to port a color scheme (Red) from VS Code to neovim and I'm wondering if this is the problem I'm having?

Here's how it looks in VS Code, note that the variable `vim` is in italics, but the members of `vim` are in a different color AND a different style (not italics).
<img width="542" height="300" alt="vscode-red-syntax-highlighting" src="https://github.com/user-attachments/assets/3fb636a3-5931-4c1d-9691-6f93836c9392" />

In the top part you can see the code from my WIP colorscheme.
Note that at (1) the yellow gets placed as intended, but it appears with italic. (it says `italic = italic` because that value is queried from the user config)
(2) and (3) show what I got from `:Inspect` (using only tree-sitter and no LSP).
<img width="736" height="489" alt="nvim-codered-hl-group-bs" src="https://github.com/user-attachments/assets/93e72823-d03a-4aa1-ba53-ae7189c2f814" />

If I change `@variable.member` to render with `bold` then that renders my `@variable.member` elements as bold and italic. So I can put on more attributes but not remove ones which are linked to?

TLDR: Given the syntax tree heirarchy, is there any way I can make vim replicate the VS Code look?

**Reply by shushtain** [reply]
Use `nocombine = true` instead of `italic = false` to disable this type of "layering".

**Reply by sasja-san** [reply]
That only sort of worked. If I declared a color value (or other attributes) on the same entity as `nocombine` then it works. But if I use `{ link = "Variable", nocombine = true, italics = true }` then it doesn't. What up with this?

**Reply by shushtain** [reply]
`link` doesn't work with any other properties; it's a full reference, without changes. Reading ":h syntax" properly will save you a lot of time.

**Reply by sasja-san** [reply]
I read it. I can't exactly say that it's clear about colors being over-writeable but attributes not being so. How was I supposed to find that out from here?

<img width="753" height="510" alt="highlight-desc" src="https://github.com/user-attachments/assets/a2a3f9bd-aacb-4b13-8c5e-a138392b41c6" />

**Reply by shushtain** [reply]
### On `italic = false` doing nothing

If you have a group "@variable" with yellow text, you don't want `DiagnosticUnderlineError` to wipe it clean, so that's why layering exists. `italic = false` doesn't cancel parent's italics the same way the absence of a color doesn't cancel parent's color. For this to happen you would use `nocombine`.

### On usage of `link`

When you use `link`, neither colors nor text/font variants can be added, as I said previously. The first paragraph from you screenshot says: you create one group to store all common styles, and you link other groups to it. It doesn't say "try adding something after link".

More than that, I bet you have found some theme to build yours upon, and at the end your table is being piped into `vim.api.nvim_set_hl()`. Sure enough, `:h nvim_set_hl()` says:

```
If `link` is used in combination with other attributes;
only the `link` will take effect (see |:hi-link|).
```

---

These questions are unrelated to the original topic of this Discussion. It's bad manners to keep pinging people from 2023 for no good reason. If you have more questions, please start a new discussion.

Also, make sure you have a nice template to start with. Leading themes (`catpuccin`, `tokyonight`, etc) use quite complex build patterns, thus they are not easy to reference. You may find [my own theme](https://github.com/shushtain/farba.nvim), `base16` ones, or some static ones more useful.

**Reply by sasja-san** [reply]
> It's bad manners to keep pinging people from 2023

You are correct. Thank you for helping me read the documentation. Ones‘ understanding is limited by how developed their web of concepts is. And when I started this, mine was pretty empty ;)

My work continues and I have now landed in way deeper trouble: <https://github.com/neovim/neovim/discussions/36657>
